httpx/middleware/basic_auth.go

```diff
@@ -0,0 +1,41 @@
+package middleware
+
+import (
+	"encoding/base64"
+	"net/http"
+	"strings"
+	"fmt"
+
+	"github.com/remind101/pkg/httpx"
+	"golang.org/x/net/context"
+)
+
+func authenticate(r *http.Request, user, pass string) bool {
+	s := strings.SplitN(r.Header.Get("Authorization"), " ", 2)
+	if len(s) != 2 || s[0] != "Basic" {
+		return false
+	}
+
+	b, err := base64.StdEncoding.DecodeString(s[1])
+	if err != nil {
+		return false
+	}
+
+	pair := strings.SplitN(string(b), ":", 2)
+	if len(pair) != 2 {
+		return false
+	}
+	return user == pair[0] && pass == pair[1]
+}
+
+func BasicAuth(h httpx.Handler, user, pass, realm string) httpx.Handler {
+	return httpx.HandlerFunc(func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
```

Comment:
What do you think about making this into a struct that implements http.Handler, like this:

```go
type BasicAuther {
        User, Pass string

        // The handler that will be called if the request is authorized.
        httpx.Handler

        // The handler that will be called if the request is not authorized. The
        // zero value is DefaultUnauthorizedHandler
        UnauthorizedHandler httpx.Handler
}

func (h *BasicAuther) ServeHTTPContext(ctx context.Context, w http.ResponseWriter, r *http.Request) error
        if authenticated(r, user, pass) {
                return h.ServeHTTPContext(ctx, w, r)
        } else {
                u := h.UnauthorizedHandler
                if u == nil {
                        u = DefaultUnauthorizedHandler
                }
                return u.ServeHTTPContext(ctx, w, r)
        }
}
```

This would give the consumer more control over the http.Handler that's called when the request is unauthorized. You can still have a `BasicAuth` factory method that returns a `BasicAuth` instance with some sensible defaults:

```go
func BasicAuth(h httpx.Handler, user, pass, realm string) *BasicAuther {
        return &BasicAuther{
                User:    user,
                Pass:    pass,
                Handler: h,
        }
}
```